File: workspace/DigitalBook/src/main/java/com/book/base/service/BookService.java
package com.book.base.service;

import java.util.List;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import com.book.base.entity.Author;
import com.book.base.entity.Book;
import com.book.base.exceptionhandler.CustomExceptionHandler;
import com.book.base.exceptionhandler.GlobleExceptionHandler;
import com.book.base.repository.IAuthorRepository;
import com.book.base.repository.IBookRepository;
import com.book.base.utility.Message;

@Service
public class BookService implements IBookService {

	@Autowired 
	private IAuthorRepository authorRepository;
	
	@Autowired
	private IBookRepository bookRepo;

	@Override
	public Book createBook(Book book) {
	   Author author =	authorRepository.findByFirstName("Micle");
	   System.out.println("Author details->"+author);
	   if(author==null)
	   {
		   throw new CustomExceptionHandler.InvalidAuthorNameException(Message.INVALID_AUTHOR);
	   }
		return bookRepo.save(book);
	}

	@Override
	public List<Book> getBooksByCategory(String category) {
		// TODO Auto-generated method stub
		List<Book> books = bookRepo.findAllByCategory(category);
		if (books.isEmpty()) {
			throw new CustomExceptionHandler.InvalidBookException(Message.EMPTY_LIST);
		}
		return books;
	}

	@Override
	public List<Book> getBooksByAuthor(String authorFirstName) {
		// TODO Auto-generated method stub
		List<Book> books=bookRepo.findAllByAuthorFirstName(authorFirstName);
		if (books.isEmpty()) {
			throw new CustomExceptionHandler.InvalidBookException(Message.EMPTY_LIST);
		}
		return books;
	}

	@Override
	public List<Book> getBooksByPrice(Double price) {
		List<Book> books= bookRepo.findAllByPrice(price);
		if (books.isEmpty()) {
			throw new CustomExceptionHandler.InvalidBookException(Message.EMPTY_LIST);
		}
		return books;
	}

	@Override
	public List<Book> getBooksByPublisher(String publisher) {

		List<Book> books= bookRepo.findAllByPublisher(publisher);
		if (books.isEmpty()) {
			throw new CustomExceptionHandler.InvalidBookException(Message.EMPTY_LIST);
		}
		return books;
	}

	@Override
	public List<Book> getAllBooks() {
		List<Book> books= bookRepo.findAll();
		if (books.isEmpty()) {
			throw new CustomExceptionHandler.InvalidBookException(Message.EMPTY_LIST);
		}
		
		return books;
	}

}

File: FRONTEND/src/app/book.service.ts
import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { Book } from './MyComponent/entity/book';

const API_URL="http://localhost:8082/digitalBooks";

@Injectable({
  providedIn: 'root'
})
export class BookService {
books:any=[];

  constructor(private httpClient: HttpClient) { }

  createBook(book:Book){
       return this.httpClient.post(API_URL+"/book",book);
  }
  getBooksByCategory(book:Book){
    const observable= this.httpClient.get(API_URL+"/books/search/category/"+book.category);
    observable.subscribe((books)=>{
     this.books=books
     
    });
    return this.books;
  }
  getBooksByAuthorName(book:Book){
    const observable= this.httpClient.get(API_URL+"/books/search/author/"+book.authorFirstName);
    observable.subscribe((books)=>{
     this.books=books 
    });
    return this.books;   
  }
  getBooksByPrice(book:Book){
    const observable= this.httpClient.get(API_URL+"/books/search/price/"+book.price);
    observable.subscribe((books)=>{
     this.books=books 
    });
    return this.books;

  }

  getBooksByPublisher(book:Book){
     const observable= this.httpClient.get(API_URL+"/books/search/publisher/"+book.publisher);
    observable.subscribe((books)=>{
     this.books=books 
    });
    return this.books;
  }

getAllBooks(){
  const observable= this.httpClient.get(API_URL+"/books/search/getAllBooks");
  observable.subscribe((books)=>{
   this.books=books 
  });
  return this.books;
}
}
